File: api_examples/graphql/clients/node/notifications-2/get-notifications-by-class/get-notifications-by-class.js
const { GraphQLClient } = require("graphql-request");
const btoa = require("btoa");

// Possible values allowed are: [ "resource", "policyValue", "policySetting", "control", "grant", "activeGrant" ]
const NOTIFICATION_TYPE_CLASS = "resource";

async function main() {
  const endpoint = process.env.TURBOT_GRAPHQL_ENDPOINT;
  const accessKeyId = process.env.TURBOT_ACCESS_KEY_ID;
  const secretAccessKey = process.env.TURBOT_SECRET_ACCESS_KEY;

  const graphQLClient = new GraphQLClient(endpoint, {
    headers: {
      authorization: "Basic " + btoa(`${accessKeyId}:${secretAccessKey}`),
    },
  });

  /* 
    The query returns unnecessary fields which will be populated depending on the value of the resource type class
    
    If the NOTIFICATION_TYPE_CLASS is set to resource, fields returned: [ resource ]
    If the NOTIFICATION_TYPE_CLASS is set to policyValue, fields returned: [ resource, policyValue ]
    If the NOTIFICATION_TYPE_CLASS is set to policySetting, fields returned: [ resource, policySetting ]
    If the NOTIFICATION_TYPE_CLASS is set to control, fields returned: [ resource, control ]
    If the NOTIFICATION_TYPE_CLASS is set to grant, fields returned: [ resource, grant ]
    If the NOTIFICATION_TYPE_CLASS is set to activeGrant, fields returned: [ resource ]
  */
  const query = `
    query {
      notifications(filter: "notificationType:${NOTIFICATION_TYPE_CLASS} limit:300 sort:-timestamp") {
        items
        {
          notificationType

          policyValue {
            default
            value
            state
            reason
            details
            secretValue
            isCalculated
          }

          policyValue {
            default
            value
            state
            reason
            details
            secretValue
            isCalculated
          }

          grant {
            permissionTypeId
            permissionLevelId
            roleName
            validFromTimestamp
            validToTimestamp
          }

          control {
            reason
            details
          }

          resource {
            object
          }
        }
      }
    }
  `;

  const variables = {};

  const data = await graphQLClient.request(query, variables);

  console.log(JSON.stringify(data, null, 2));
}

main().catch((error) => console.error(error));

File: guardrails_utilities/python_utils/notifications/get_notifications_by_class/get-notifications-by-class.py
import json
import click
import turbot
from sgqlc.endpoint.http import HTTPEndpoint

# -c .config/turbot/credentials.yml --profile env --notification_class resource
@click.command()
@click.option('-c', '--config-file', type=click.Path(dir_okay=False), help="[String] Pass an optional yaml config file.")
@click.option('-p', '--profile', default="default", help="[String] Profile to be used from config file.")
# Possible values allowed are: [ "resource", "policyValue", "policySetting", "control", "grant", "activeGrant" ]
@click.option('-c', '--notification_class', default="resource", help="[String] Set the notification class, for more information see https://turbot.com/v5/docs/reference/filter/notifications#filtering-notifications.")
@click.option('-d', '--datetime_filter', default="<T-10d", help="[String] Configures the date range to filter the result, for more information see https://turbot.com/v5/docs/reference/filter#datetime-filters.")
@click.option('-s', '--sort', default="-timestamp", help="[String] The field to use for sorting the results, for more information see https://turbot.com/v5/docs/reference/filter#sorting.")
def run_controls(config_file, profile, notification_class, datetime_filter, sort):
    config = turbot.Config(config_file, profile)
    headers = {'Authorization': 'Basic {}'.format(config.auth_token)}
    endpoint = HTTPEndpoint(config.graphql_endpoint, headers)

    # The query returns unnecessary fields which will be populated depending on the value of the resource type class
    #
    # If the type is a member of the notification class resource, fields returned: [ resource ]
    # If the type is a member of the notification class policyValue, fields returned: [ resource, policyValue ]
    # If the type is a member of the notification class policySetting, fields returned: [ resource, policySetting ]
    # If the type is a member of the notification class control, fields returned: [ resource, control ]
    # If the type is a member of the notification class grant, fields returned: [ resource, grant ]
    # If the type is a member of the notification class activeGrant, fields returned: [ resource ]
    #
    query = '''
      query {
        notifications(filter: "notificationType:%s timestamp:%s sort:%s") {
          items
          {
            notificationType

            policyValue {
              default
              value
              state
              reason
              details
              secretValue
              isCalculated
            }

            policyValue {
              default
              value
              state
              reason
              details
              secretValue
              isCalculated
            }
            grant {
              permissionTypeId
              permissionLevelId
              roleName
              validFromTimestamp
              validToTimestamp
            }

            control {
              reason
              details
            }

            resource {
              object
            }
          }
        }
      }
    ''' % (notification_class, datetime_filter, sort)

    print(f'Query notifications for type class: {notification_class}')
    print(f'Applying timestamp filter: {datetime_filter}')
    print(f'Applying sorting: {sort}')

    empty_variables = {}
    query_response = endpoint(query, empty_variables)
    if "errors" in query_response:
        return report_graphql_error("aws_mutation", query_response["errors"])

    for item in query_response["data"]["notifications"]["items"]:
        print(json.dumps(item, indent=1))


def report_graphql_error(section, errors):
    print("Error(s) detected when running mutation: {}".format(section))
    for error in errors:
        print(error["message"])


if __name__ == "__main__":
    run_controls()
